syntax = "proto3";

import "google_protobuf_wrappers_cp.proto";

message Thresholds {
  google_cp.protobuf.DoubleValue warning = 1;  // empty value means it's not applicable.
  google_cp.protobuf.DoubleValue failure = 2;  // empty value means it's not applicable.
}

enum CheckStatisticType {
  CHECK_STATISTIC_TYPE_UNDEFINED = 0;
  CHECK_STATISTIC_TYPE_ECL_ESTIMATOR_FAILURE_MAX = 1;
  CHECK_STATISTIC_TYPE_ECL_ESTIMATOR_FAILURE_AVG = 2;
  CHECK_STATISTIC_TYPE_ECL_ESTIMATOR_FAILURE_WINDOWED_AVG = 3;
  CHECK_STATISTIC_TYPE_ECL_INNOVATION_AMBER_PCT = 4;
  CHECK_STATISTIC_TYPE_ECL_INNOVATION_AMBER_WINDOWED_PCT = 5;
  CHECK_STATISTIC_TYPE_ECL_INNOVATION_RED_PCT = 6;
  CHECK_STATISTIC_TYPE_ECL_INNOVATION_RED_WINDOWED_PCT = 7;
  CHECK_STATISTIC_TYPE_ECL_FAIL_RATIO_PCT = 8;
  CHECK_STATISTIC_TYPE_ECL_FAIL_RATIO_WINDOWED_PCT = 9;
  CHECK_STATISTIC_TYPE_ECL_IMU_CONING_MAX = 10;
  CHECK_STATISTIC_TYPE_ECL_IMU_CONING_AVG = 11;
  CHECK_STATISTIC_TYPE_ECL_IMU_CONING_WINDOWED_AVG = 12;
  CHECK_STATISTIC_TYPE_ECL_IMU_HIGH_FREQ_DELTA_ANGLE_MAX = 13;
  CHECK_STATISTIC_TYPE_ECL_IMU_HIGH_FREQ_DELTA_ANGLE_AVG = 14;
  CHECK_STATISTIC_TYPE_ECL_IMU_HIGH_FREQ_DELTA_ANGLE_WINDOWED_AVG = 15;
  CHECK_STATISTIC_TYPE_ECL_IMU_HIGH_FREQ_DELTA_VELOCITY_MAX = 16;
  CHECK_STATISTIC_TYPE_ECL_IMU_HIGH_FREQ_DELTA_VELOCITY_AVG = 17;
  CHECK_STATISTIC_TYPE_ECL_IMU_HIGH_FREQ_DELTA_VELOCITY_WINDOWED_AVG = 18;
  CHECK_STATISTIC_TYPE_ECL_IMU_OBSERVED_ANGLE_ERROR_WINDOWED_AVG = 19;
  CHECK_STATISTIC_TYPE_ECL_IMU_OBSERVED_VELOCITY_ERROR_WINDOWED_AVG = 20;
  CHECK_STATISTIC_TYPE_ECL_IMU_OBSERVED_POSITION_ERROR_WINDOWED_AVG = 21;
  CHECK_STATISTIC_TYPE_ECL_IMU_DELTA_ANGLE_BIAS_AVG = 22;
  CHECK_STATISTIC_TYPE_ECL_IMU_DELTA_ANGLE_BIAS_WINDOWED_AVG = 23;
  CHECK_STATISTIC_TYPE_ECL_IMU_DELTA_VELOCITY_BIAS_AVG = 24;
  CHECK_STATISTIC_TYPE_ECL_IMU_DELTA_VELOCITY_BIAS_WINDOWED_AVG = 25;
}

message CheckStatistic {
  CheckStatisticType type = 1;
  double value = 2;
  Thresholds thresholds = 3;
}

enum CheckStatus {
  CHECK_STATUS_UNDEFINED = 0;
  CHECK_STATUS_PASS = 1;
  CHECK_STATUS_WARNING = 2;
  CHECK_STATUS_FAIL = 3;
  CHECK_STATUS_DOES_NOT_APPLY = 4;  // E.g. No VIO; optical flow sensor check is not meaningful.
}

enum CheckType {
  CHECK_TYPE_UNDEFINED = 0;
  CHECK_TYPE_ECL_MAGNETOMETER_STATUS = 1;
  CHECK_TYPE_ECL_MAGNETIC_HEADING_STATUS = 2;
  CHECK_TYPE_ECL_VELOCITY_SENSOR_STATUS = 3;
  CHECK_TYPE_ECL_POSITION_SENSOR_STATUS = 4;
  CHECK_TYPE_ECL_HEIGHT_SENSOR_STATUS = 5;
  CHECK_TYPE_ECL_HEIGHT_ABOVE_GROUND_SENSOR_STATUS = 6;
  CHECK_TYPE_ECL_AIRSPEED_SENSOR_STATUS = 7;
  CHECK_TYPE_ECL_SIDESLIP_SENSOR_STATUS = 8;
  CHECK_TYPE_ECL_IMU_VIBRATION_STATUS = 9;
  CHECK_TYPE_ECL_IMU_BIAS_STATUS = 10;
  CHECK_TYPE_ECL_IMU_OUTPUT_PREDICTOR_STATUS = 11;
  CHECK_TYPE_ECL_OPTICAL_FLOW_STATUS = 12;
  CHECK_TYPE_ECL_FILTER_FAULT_STATUS = 13;
}

message CheckResult {
  CheckStatus status = 1;
  CheckType type = 2;
  repeated CheckStatistic statistics = 3;
}
